In [1]:
import pandas as pd 
import numpy as np 

In [2]:
data = pd.read_csv("/Users/sarthaktyagi/Desktop/projects /fraudDetection/data/PS_20174392719_1491204439457_log.csv")

In [3]:
data.shape

(6362620, 11)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [7]:
data.sample(5)

step      type     amount     nameOrig  oldbalanceOrg  \
4684554   331  CASH_OUT  158245.15  C1828268654           0.00   
4133847   303   PAYMENT    6375.32  C1828350013      115846.89   
5066896   354   PAYMENT    6280.43  C1396983985        5879.00   
2931760   229  CASH_OUT  314739.74  C1027535010       19459.00   
1901796   165   PAYMENT    3496.84  C1753011794        3487.00   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
4684554            0.00   C621284533      1817899.92      1976145.07        0   
4133847       109471.57  M1465411239            0.00            0.00        0   
5066896            0.00  M1204928344            0.00            0.00        0   
2931760            0.00   C627627530        23193.24       337932.98        0   
1901796            0.00  M1097712392            0.00            0.00        0   

         isFlaggedFraud  
4684554               0  
4133847               0  
5066896               0  
2931760               0  
1901796               0

In [10]:
data['isFraud'].value_counts()

isFraud
0    6354407
1       8213
Name: count, dtype: int64

In [12]:
data = data.drop(columns=['isFlaggedFraud'],axis=1)

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 10 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
dtypes: float64(5), int64(2), object(3)
memory usage: 485.4+ MB


In [14]:
data.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
dtype: int64

In [15]:
data.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  6.362620e+06  6.362620e+06   6.362620e+06    6.362620e+06   
mean   2.433972e+02  1.798619e+05   8.338831e+05    8.551137e+05   
std    1.423320e+02  6.038582e+05   2.888243e+06    2.924049e+06   
min    1.000000e+00  0.000000e+00   0.000000e+00    0.000000e+00   
25%    1.560000e+02  1.338957e+04   0.000000e+00    0.000000e+00   
50%    2.390000e+02  7.487194e+04   1.420800e+04    0.000000e+00   
75%    3.350000e+02  2.087215e+05   1.073152e+05    1.442584e+05   
max    7.430000e+02  9.244552e+07   5.958504e+07    4.958504e+07   

       oldbalanceDest  newbalanceDest       isFraud  
count    6.362620e+06    6.362620e+06  6.362620e+06  
mean     1.100702e+06    1.224996e+06  1.290820e-03  
std      3.399180e+06    3.674129e+06  3.590480e-02  
min      0.000000e+00    0.000000e+00  0.000000e+00  
25%      0.000000e+00    0.000000e+00  0.000000e+00  
50%      1.327057e+05    2.146614e+05  0.000000e+00  
75%      9.430367e+05    1.111909e+06  0.000000e+00  
max      3.560159e+08    3.561793e+08  1.000000e+00

In [21]:
data.duplicated().sum()

np.int64(0)

In [24]:
data.columns

Index(['step', 'type', 'amount', 'nameOrig', 'oldbalanceOrg', 'newbalanceOrig',
       'nameDest', 'oldbalanceDest', 'newbalanceDest', 'isFraud'],
      dtype='object')

##### intialily there was 11 columns and 6362620 rows
##### columns =['step', 'type', 'amount', 'nameOrig', 'oldbalanceOrg', 'newbalanceOrig','nameDest', 'oldbalanceDest', 'newbalanceDest', 'isFraud','isflagedFruad']
and clearly we don't need the last feature "isflagedFruad"

## dataset insights
1. data is highly imbalanced with value counts 
    isFraud
    0-->6354407
    1-->8213

2. there is no duplicate and null values 


# feature engineering 


In [25]:
# removing the less important columns 
data = data.drop(columns=['nameOrig','nameDest'],axis=1)

In [40]:
categorical_columns =[]
numerical_columns =[]

for column in data.columns:
    if data[column].dtypes == object:
        categorical_columns.append(column)
    else:
        numerical_columns.append(column)
    

In [41]:
categorical_columns

['type']

In [42]:
numerical_columns

['step',
 'amount',
 'oldbalanceOrg',
 'newbalanceOrig',
 'oldbalanceDest',
 'newbalanceDest',
 'isFraud']

In [46]:
# conveting categorcal values to numerical (ohe)
newData = pd.get_dummies(data,columns=categorical_columns,drop_first=True)

In [47]:
newData.shape

(6362620, 11)

In [48]:
newData.sample(4)

step     amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
4610775   329  157298.49     5847087.75      6004386.24       523366.60   
5394174   377   44920.82     4835048.94      4879969.76      1985784.65   
5945052   405  353278.17        1666.00       354944.17         4433.88   
2641785   209  172194.14       15441.00            0.00            0.00   

         newbalanceDest  isFraud  type_CASH_OUT  type_DEBIT  type_PAYMENT  \
4610775       366068.11        0          False       False         False   
5394174      1940863.82        0          False       False         False   
5945052            0.00        0          False       False         False   
2641785       172194.14        0           True       False         False   

         type_TRANSFER  
4610775          False  
5394174          False  
5945052          False  
2641785          False

In [51]:
newData['errorBalanceOrig'] = newData['newbalanceOrig'] + newData['amount'] - newData['oldbalanceOrg']
newData['errorBalanceDest'] = newData['oldbalanceDest'] + newData['amount'] - newData['newbalanceDest']


In [53]:
newData[newData['isFraud'] == 0]

step     amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0           1    9839.64       170136.0       160296.36            0.00   
1           1    1864.28        21249.0        19384.72            0.00   
4           1   11668.14        41554.0        29885.86            0.00   
5           1    7817.71        53860.0        46042.29            0.00   
6           1    7107.77       183195.0       176087.23            0.00   
...       ...        ...            ...             ...             ...   
6362319   718    8634.29       518802.0       510167.71            0.00   
6362320   718  159188.22         3859.0            0.00            0.00   
6362321   718  186273.84       168046.0            0.00        24893.67   
6362322   718   82096.45        13492.0            0.00            0.00   
6362323   718    1864.24        20426.0        18561.76       188746.00   

         newbalanceDest  isFraud  type_CASH_OUT  type_DEBIT  type_PAYMENT  \
0                  0.00        0          False       False          True   
1                  0.00        0          False       False          True   
4                  0.00        0          False       False          True   
5                  0.00        0          False       False          True   
6                  0.00        0          False       False          True   
...                 ...      ...            ...         ...           ...   
6362319            0.00        0          False       False          True   
6362320       159188.22        0           True       False         False   
6362321       211167.51        0           True       False         False   
6362322        82096.45        0          False       False         False   
6362323       190610.24        0          False        True         False   

         type_TRANSFER  errorBalanceOrig  errorBalanceDest  
0                False              0.00           9839.64  
1                False              0.00           1864.28  
4                False              0.00          11668.14  
5                False              0.00           7817.71  
6                False              0.00           7107.77  
...                ...               ...               ...  
6362319          False              0.00           8634.29  
6362320          False         155329.22              0.00  
6362321          False          18227.84              0.00  
6362322           True          68604.45              0.00  
6362323          False              0.00              0.00  

[6354407 rows x 13 columns]

In [54]:
newData.isFraud.value_counts()

isFraud
0    6354407
1       8213
Name: count, dtype: int64